theano/sandbox/cuda/opt.py

```diff
@@ -2313,3 +2313,73 @@ def local_gpu_allocempty(node):
+                    if k not in ['mode']:
+                        return
+            # We make it simple, we suppose all float32 inputs will be
+            # on the GPU and that all outputs in float32 will be on
+            # the GPU.
+            # We do not want to add the shared inputs. They will be computed again.
+            new_outer_inp = []  # new outer inputs.
+            inner_replace = {}
+            for o_inp, i_inp in zip(node.inputs, node.op.inputs):
+                if getattr(o_inp, 'dtype', None) == 'float32':
+                    new_outer_inp.append(as_cuda_ndarray_variable(o_inp))
+                    inner_replace[i_inp] = tensor.as_tensor_variable(
+                        CudaNdarrayType(dtype=i_inp.dtype,
+                                        broadcastable=i_inp.broadcastable)())
+                else:
+                    new_outer_inp.append(inp)
+            new_inner_out = []
+            for out in node.op.outputs:
+                if getattr(out, 'dtype', None) == 'float32':
+                    new = as_cuda_ndarray_variable(out)
+                    new_inner_out.append(new)
+                else:
+                    new_inner_out.append(out)
+
+            # SharedVariable was cloned when the FunctionGraph was
+            # created.  node.op.inputs still make reference to the
+            # original SharedVariable.  We must swap it to the one
+            # used in this fgraph
+            for i in range(len(node.op.shared_inputs)):
+                # SharedVariable the user used
+                orig_shr = node.op.shared_inputs[i]
+                # SharedVariable after cloning orig_shr when
+                # FunctionGraph cloned the user graph.
+                fgraph_shr = node.inputs[len(node.op.inputs) + i]
+                inner_replace[orig_shr] = fgraph_shr
+            # rebuild the inner graph to introduce the gup_from_host
+            # at the start.
+            new = theano.compile.rebuild_collect_shared(
+                new_inner_out, inputs=node.op.inputs,
+                replace=inner_replace,
+                copy_inputs_over=True)
+            (new_inputs, new_outputs, _) = new
+            # new_inputs give us the corresponding to the old inputs.
+            for i in range(len(new_inputs)):
+                if (new_inputs[i].owner and
+                        isinstance(new_inputs[i].owner.op, HostFromGpu)):
+                    new_inputs[i] = new_inputs[i].owner.inputs[0]
+            op = theano.compile.builders.OpFromGraph(
+                new_inputs, new_outputs, **node.op.kwargs)
+            ret = op(*new_outer_inp, **dict(return_list=True))
+            assert all([inp in node.fgraph.variables for inp in theano.gof.graph.inputs(ret)])
+
+            for i in range(len(ret)):
+                if getattr(node.op.outputs[i], 'dtype', None) == 'float32':
+                    ret[i] = theano.tensor.as_tensor_variable(ret[i])
+            return ret
```

Comment:
I went through the code(not the test) just now.  Following is my idea, please correct my mistake. I don't want to start on a wrong direction:

1. It seems this code convert every possible inputs&output into ```CudaNdarray```. 
2. Therefore in the next step ```rebuild_collection_shared()```, ```variable.filter_variable()``` will introduce the ```GpuFromHost``` OP. 
3.  Next, the code contains the ```HostFromGpu``` in front of the inputs. 
4. Finally it use the new inputs&outputs to init a new OpFromGraph, the Optimization process within the inner_function will therefore delete unnecessary copy of data.

-----------------------------------

So, is there any missing feature I need to implement? 
I saw there's a note ```Need to implement updates, givens, ...```, that mean we need to do something to updates&givens? I am not very familiar with optimization,  so you could show some "user case" while explaining to me what I need to handle.